SyncLink impls Links, minor changes to Link API and BlackHole

ddowl opened the pull request:
Commit messages are pretty descriptive, but this PR takes the next leap to have `SyncLink` impl `Link`, or rather a `SyncLinkBuilder` impl `LinkBuilder`. It's up for discussion, but I think having an explicit `*Builder` makes sense since the purpose of these objects are now to keep track of enough inputs to construct some `Link`, which is now defined as the product type of the necessary `TokioRunnables` and `PacketStream<Output>` (i.e. egressors) that need to be driven and exposed, respectively. The name `Link` in particular is confusing because of its type declaration `Link<Output>`, which isn't totally clear. I've also disabled `graphgen` tests.

ddowl committed:
SyncLink: impl Link

In addition, prefers the name "in_streams" since it's synonymous with
"input_streams" and "ingress_streams". Removes graphgen examples from
top-level crate so they aren't compiled while API is in flux. Modifies
the `ingressor` function to consume `self`, as builder methods always
yield a new version of `self`. This is useful when moving the Link to
a new state (see `SyncLink::ingressors` and `SyncLink::element`).

ddowl committed:
BlackHoleLink: build BlackHole on build_link

ddowl committed:
Link: rename to LinkBuilder, create Link product type

ddowl commented on route-rs-runtime/src/link/mod.rs:
```diff
@@ -31,25 +31,35 @@ pub use self::tee_link::*;
 /// Some Links may need to be driven by Tokio. This represents a handle to something Tokio can run.
 pub type TokioRunnable = Box<dyn Future<Item = (), Error = ()> + 'static + Send>;
+/// LinkBuilders build this.
+pub type Link<Output> = (Vec<TokioRunnable>, Vec<PacketStream<Output>>);
```
Better name pls?

CollinValley replied:
I mean, it isn't the worst name right now.  LinkBuilder's build Links, that is pretty clear? Still open to ideas though.

ddowl commented on route-rs-runtime/src/link/sync_link.rs:
```diff
@@ -1,22 +1,73 @@
+
+        SyncLinkBuilder {
+            in_streams: Some(in_streams),
+            element: self.element,
```
I think this pattern of builder methods passing all of `self`'s state except for the input is going to become common. I'm not sure if anything can be done to abstract that idea, but we should wait until every `Link` is migrated before trying.

ddowl commented on route-rs-runtime/src/link/sync_link.rs:
```diff
@@ -1,22 +1,73 @@
+/// storage. In the future we might decide to restrict the interface for this link
+/// for clearer intent.
+impl<E: Element + Send + 'static> LinkBuilder<E::Input, E::Output> for SyncLinkBuilder<E> {
+    fn ingressors(self, in_streams: Vec<PacketStream<E::Input>>) -> Self {
```
Do we want to enforce that each builder method should be called at most once? For example, we could assert here that `self.in_streams` must be `None`, otherwise panic with a helpful message.

CollinValley replied:
I think the question is: Does running the function twice cause issues other than the obvious  'The values change' answer.  If yes, then we should prevent double running.

ddowl replied:
I mean, no it shouldn't, but we should be able to quickly detect nonsensical builders like:

```
SyncLinkBuilder
  .ingressors(<some vec>)
  .ingressors(<wait... another vec??>)
  .element(e)
  .build_link();
```

I guess running the function twice causes the issue of "being confusing to the reader". Functionally the last builder value is used.

CollinValley replied:
Hmm, if you want to make a double call panic, I am fine with that.

ddowl commented on route-rs-runtime/src/link/sync_link.rs:
```diff
@@ -1,22 +1,73 @@
+}
+
+/// The single egressor of SyncLink
+struct SyncProcessor<E: Element> {
```
The idea behind "Processor" is that this is the thing that performs the transformation based on the `Link`'s `Element::process`, rather than just a handle for outgoing packets. It's a little strange since this entity is semantically different in different contexts: inside `SyncLink` it's a packet processor, but outside `SyncLink` it's an upstream `PacketStream` handle.

CollinValley replied:
I like the processor name, it tells us what it does in the code. And to the user, they never see the name. They only really interact with it through the TokioRunnables and Egressors. Seems like a good name to me.

ddowl commented on route-rs-runtime/src/link/blackhole_link.rs:
```diff
@@ -1,41 +1,44 @@
+impl<Packet: Sized + 'static> LinkBuilder<Packet, ()> for BlackHoleLinkBuilder<Packet> {
+    fn ingressors(self, ingress_streams: Vec<PacketStream<Packet>>) -> Self {
+        BlackHoleLinkBuilder {
+            in_streams: Some(ingress_streams),
```
With multiple builder functions, the pattern of "accumulate all of your state, then initialize in `build_link`" became clearer, so `BlackHoleLink` is updated to reflect that.

CollinValley commented on route-rs-runtime/src/link/sync_link.rs:
```diff
@@ -42,13 +93,15 @@ impl<E: Element> Stream for SyncLink<E> {
-                    // if `element.process` returns None, do nothing, loop around and try polling again.
-                    if let Some(output_packet) = self.element.process(input_packet) {
-                        return Ok(Async::Ready(Some(output_packet)));
+            for ingress_stream in self.in_streams.iter_mut() {
```
Issue here: Since we are servicing multiple inputs, we need to fairly pull from them. If you look at join link, you want similar functionality to that.  The trick is usually to just save the index of the last thing you pulled a packet form internally, then restart the for loop at that index on the next poll.

CollinValley replied:
You probably want something along the lines of:
```
    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
        //rotate_slice exists in 1.22 nightly experimental
        let rotated_iter = self
            .from_ingressors
            .iter()
            .enumerate()
            .cycle()
            .skip(self.next_pull_ingressor)
            .take(self.from_ingressors.len());
        for (port, from_ingressor) in rotated_iter {
            match from_ingressor.try_recv() {
                Ok(Some(packet)) => {
                    unpark_and_notify(&self.task_parks[port]);
                    self.next_pull_ingressor = port + 1;
                    return Ok(Async::Ready(Some(packet)));
                }
                Ok(None) => {
                    //Got a none from a consumer that has shutdown
                    self.ingressors_alive -= 1;
                    if self.ingressors_alive == 0 {
                        return Ok(Async::Ready(None));
                    }
                }
                Err(_) => {
                    //On an error go to next channel.
                }
            }
        }
```

ddowl replied:
Offline, we spent a meetup figuring out that fairly polling many ingress streams requires some kind of intermediary channel and task_park like `JoinLink` has. This is the reason why we decided `SyncLink`s should be 1 to 1 rather than n to 1.

Since `AsyncLink` will need to fairly poll its ingress streams, fair joining functionality can be implemented there.

CollinValley reviewed (changes requested):
Overall, the actual changes look really good. With the for loop functionality fixed I think this should be ready for merge.

ddowl commented:
Before merging this, I'd like to make a few changes to `SyncLink` to simplify polling logic since we are guaranteed to only have 1 ingress stream.

ddowl committed:
SyncLink: enforce 1 ingress stream

Changes the SyncLink* structs to store only 1 ingress stream, and adds a
convenient helper constructor that takes only 1 PacketStream as opposed
to a Vec<PacketStream>.

CollinValley reviewed (approved):
LGTMMD
